### ISSUE TITLE
Allow `func-names` to not error for missing, but inferrable names

### ISSUE THREAD

**Issue by loganfsmyth** [eslint/eslint#7235]
The `func-names` rule is very aggressive at the moment and will error on any function expression with no name. While useful, in an ES6 environment there are now many places where the name will be automatically added by the execution environment, but where this rule will see an error.

Examples of current errors:

```
var foo = function(){};
foo = function(){};
[foo = function(){}] = [];
export default (function(){});
```

The rule has an `always` and `never` option right now. I'd propose adding an `"uninferred"` option which would be an alternative to `"always"` that would not error for the cases where ES6 knows the name.

I've played around locally, so I can PR this, but it seems like policy is to discuss here first?

**Comment by platinumazure** [comment]
I'm pretty sure this is the whole purpose of the rule: warn when a function expression lacks a name, even when it could be inferred. It doesn't warn on function declarations, because those require a name due to the ECMAScript grammar. So if you don't want those cases to be flagged, then simply don't use the rule.

Are there any examples you haven't listed, but for which you believe this option would be necessary? In other words, a function expression with a "non-inferrable name" which would need to be enforced using your proposed option?

Also: Please ensure you have filled out everything in the Propose Rule Change template, linked in the below post from eslintbot. Thanks!

**Comment by eslintbot** [comment]
Hi @loganfsmyth, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by loganfsmyth** [comment]
Perhaps I was making off-base assumptions about this rule. My understanding of the objective of it is "make sure all functions have names in stack traces".

> Are there any examples you haven't listed, but for which you believe this option would be necessary? In other words, a function expression with a "non-inferrable name" which would need to be enforced using your proposed option?

Totally. Given the goal of making stacktraces easy, inferred names aren't needed, but anywhere they won't be inferred would still need a name, which could be things like

```
aFunction(function(){});
obj.foo = function(){};
```

would still need a name since none will be inferred as far as I know, and it would show as "anonymous" in the stack trace.


**What rule do you want to change?**

`func-names`

**Does this change cause the rule to produce more or fewer warnings?**

Fewer, when `uninferred` is in use rather than `always`.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option

**Please provide some example code that this change will affect:**

```js
var foo = function(){};
```

**What does the rule currently do for this code?**

It will error because the function has no name. While true for ES5, in ES6 the function will be given a name based on the variable it is assigned to.

**What will the rule do after it's changed?**

If `uninferred` is used instead, it will not recognize this as an error.

**Comment by loganfsmyth** [comment]
Also FYI the `rule change proposal template` link on http://eslint.org/docs/developer-guide/contributing/rule-changes does not work, it is missing the `.md` extension for some reason

**Comment by ilyavolodin** [comment]
@loganfsmyth We don't really have rules that branch reporting logic based on the version of ECMAScript you are enabling. I know that this change would, for example, be very hard to consume for me, since I have Rhino code that is still ES5 but all tests are ES6, and I want to force naming for all functions. At most, I think we could consider an opt-in option for this rule to disable enforcing names when it can be inferred.

**Comment by loganfsmyth** [comment]
> At most, I think we could consider an opt-in option for this rule to disable enforcing names when it can be inferred.

Totally, that's what I'm proposing. To have `always`, `uninferred`, and `never` be three options for `func-names` instead of just the current two.

**Comment by ilyavolodin** [comment]
Hmm.. OK, I can buy the need for the `uninferred` option. I'll support it. 👍

**Comment by platinumazure** [comment]
As long as we don't actually detect ES6 as part of the new option (i.e., we just assume ES6 and note in the documentation that "uninferred" is according to ES6 rules), I can get behind this.

**Comment by loganfsmyth** [comment]
Cool, PR: https://github.com/eslint/eslint/pull/7244 Feedback welcome!

**Comment by kaicataldo** [comment]
I'll support this change. One thing to note - we have a PR open right now for a new rule that does actually detect the `ecmaVersion` being used: https://github.com/eslint/eslint/pull/7063#issuecomment-247884530. Following that pattern, we could actually potentially modify the "always" option to do a check for ES6+ in the rule (defaulting to ES5 if not present).

**Comment by not-an-aardvark** [comment]
I like this idea, but I'm not a big fan of the option name `uninferred`. In my opinion, that name doesn't describe the idea very well; it's a type of function, rather than an explicit preference for whether that function should have a name.

Maybe something like `as-needed` would be better? That would make the option names consistent with `quote-props`, `arrow-body-style`, and `arrow-parens`, and I think it makes it clearer that a name is only required if no name would be inferred.

**Comment by loganfsmyth** [comment]
I don't have strong feelings about the name really, my initial concern was that `as-needed` didn't give enough info, but depending on your ES6 knowledge I guess that's the case with `uninferred` too. Definitely up for changing.

**Comment by nzakas** [comment]
I'm against this change. The rule is strictly about whether a function has a formal name defined, not whether the name property has a value. ES6 specifically provides that all functions have a valid value for the name property, so this would amount to turning the rule off. 

I don't think it makes sense to use this rule if you only care about the name property of functions. I think a new custom rule is more appropriate than changing the meaning of this rule.

**Comment by loganfsmyth** [comment]
> The rule is strictly about whether a function has a formal name defined, not whether the name property has a value.

Would you be up for elaborating on your reasoning behind that? Until ES6, those were one and the same anyway, but it seems like ES6 name inference opens that assumption up for reassessment. It feels like the core of my question here comes down to the definition of `has a formal name defined`, since to me `var foo = function(){}` does have a formal name defined.

> ES6 specifically provides that all functions have a valid value for the name property

That's not true for cases like `callWithCallback(function(){   });` or `obj.prop = function(){}`.

**Comment by nzakas** [comment]
@loganfsmyth let's not get into debates over semantics (I was typing on a phone, so I try to make things as concise as possible). Speaking in terms of the AST, the original point of the rule was to make sure that there's always an `id` on a `FunctionExpression` node. Literally, that's [all it does](https://github.com/eslint/eslint/blob/master/lib/rules/func-names.js#L66).

What you're asking for is to only warn when the `name` property of a function would be `""`. To me, that seems like a different use case.

If other team members feel strongly about this, I'll withdraw my objection - it just seems a bit out of place to me, but I could be wrong.

**Comment by loganfsmyth** [comment]
> Speaking in terms of the AST, the original point of the rule was to make sure that there's always an id on a FunctionExpression node.

Sounds good. I think I just made an incorrect assumption about the overall goal of this rule. Our usecase of "always" has generally been for debugging, and I ended up associating this rule with `.name`. Looking at it from the other side, I can see that `never` for instance wouldn't make sense as an option for a `.name`-based rule.

Should I pursue this as a entirely custom rule, or do you think a rule like `func-name-property` rule with `es5` and `es6` would be something worthwhile?

**Comment by nzakas** [comment]
I'm going to bring this up at the TSC meeting to get more clarity.

**Comment by btmills** [comment]
After discussion in the 2016-10-13 TSC meeting, we decided to accept this enhancement. A PR that closes this issue should also clarify in the rule's docs that the rule's aim is to ensure names are present for debugging purposes.

**Comment by ValYouW** [comment]
Hi,
Excuse me for bringing this issue up, but why `Foo.prototype.bar = function() {};` is incorrect when using the `as-needed` option? while I understand that the function's name property would be empty, in callstacks this function will have the name `Foo.bar`, and I think this goes along with the TSC decision of "see the rule as ensuring names are present for debugging".
Also this is the behavior of JSCS `requireNamedUnassignedFunctions` (which is actually what I'm trying to restore)...
Thx.

**Comment by kaicataldo** [comment]
@ValYouW Mind making a new issue for this so that we can discuss? This won't get much visibility here. Thanks!

**Comment by ValYouW** [comment]
@kaicataldo Done. #7739

**Comment by kaicataldo** [comment]
Thanks!
